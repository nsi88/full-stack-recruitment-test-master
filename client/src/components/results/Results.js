import React from 'react';
import './Results.scss';
// TODO: use one arrow icon passing a color as a prop.
import longArrowRightGray from '../../icons/long-arrow-right-gray.svg';
import dateTimeHelper from '../../date-time-helper.js';

const Results = ({results, fetchingInProgress}) => {
  if (fetchingInProgress) {
    return (
      <div className='spinner'>
        <div className='spinner__icon'/>
        <div className='spinner__text'>Searching</div>
      </div>
    );
  }

  return (
    <ul className='results'>
      {results.map((result) => (<Result result={result}/>))}
    </ul>
  );
};

const Result = ({result}) => (
  <li className='result'>
    <div className='result__legs'>
      {result.Legs.map((leg) => (<Leg leg={leg}/>))}
    </div>
    <div className='result__bottom'>
      <div className='result__price-info'>
        <div className='result__price'>
          {result.Currency.Symbol}&nbsp;{result.Price}
        </div>
        <div className='result__agent'>{result.Agent.Name}</div>
      </div>
      <button className='result__select'>Select</button>
    </div>
  </li>
);

const Leg = ({leg}) => (
  <div className='leg'>
    <div className='leg__main-info'>
      <img className='leg__airline' alt={leg.Carrier.Name} src={carrierLogo(leg.Carrier.Code)}/>
      <div className='leg__departure'>
        <div className='leg__time'>{dateTimeHelper.formatTime(dateTimeHelper.parse(leg.Departure))}</div>
        <div className='leg__place'>{leg.OriginStation.Code}</div>
      </div>
      <div className='leg__arrow-container'>
        <img className='leg__arrow' alt='to' src={longArrowRightGray}/>
      </div>
      <div className='leg__arrival'>
        <div className='leg__time'>{dateTimeHelper.formatTime(dateTimeHelper.parse(leg.Arrival))}</div>
        <div className='leg__place'>{leg.DestinationStation.Code}</div>
      </div>
    </div>
    <div className='leg__additional-info'>
      <div className='leg__duration'>{dateTimeHelper.formatDuration(leg.Duration)}</div>
      <div className={leg.StopsCount === 0 ? 'leg__stops leg--direct' : 'leg__stops leg--with-stops'}>{stopsDescription(leg.StopsCount)}</div>
    </div>
  </div>
);


/**
 * stopsDescription - Text description of stops count.
 *
 * E.g. 0 -> 'Direct', 1 -> '1 stop', 2 -> '2 stops'.
 */
function stopsDescription(stopsCount) {
  switch (stopsCount) {
    case 0:
      return 'Direct';
    case 1:
      return '1 stop';
    default:
      return `${stopsCount} stops`
  }
}

// REVIEW: not sure in here, do I need to take imageUrl or make ut mannually, like now.
function carrierLogo(carrierCode) {
  return `https://logos.skyscnr.com/images/airlines/favicon/${carrierCode}.png`;
}

export default Results;
