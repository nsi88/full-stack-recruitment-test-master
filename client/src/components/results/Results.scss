@import "~bpk-mixins";

.results {
  list-style-type: none;
  margin: 0;
  padding: 0;
  display: flex;
  flex-direction: column;
  background-color: $bpk-color-gray-50;
}

.result {
  display: flex;
  flex-direction: column;
  margin: $bpk-spacing-sm $bpk-spacing-sm $bpk-spacing-xs;
  padding: $bpk-spacing-xs $bpk-spacing-sm 0;
  background-color: $bpk-color-white;
  @include bpk-box-shadow-lg;
  @include bpk-border-radius-sm;

  &__bottom {
    display: flex;
    justify-content: space-between;
    align-items: center;
  }

  &__price-info {
    padding: $bpk-spacing-xs 0 0;
  }

  &__price {
    @include bpk-text-xl;
    color: $bpk-color-gray-900;
  }

  &__agent {
    @include bpk-text-base;
    color: $bpk-color-gray-300;
    position: relative;
    top: -0.75rem;
  }

  &__select {
    @include bpk-button;
    margin: $bpk-spacing-xs 0 0;
    padding: $bpk-spacing-xs $bpk-spacing-lg;
  }
}

.leg {
  display: flex;
  justify-content: space-between;
  margin: 0 0 $bpk-spacing-sm;

  &:last-child {
    margin-bottom: 0;
  }

  &__main-info {
    display: flex;
    align-items: center;
  }

  &__additional-info {
    @include bpk-text-sm;
    display: flex;
    flex-direction: column;
    align-content: space-between;
    padding: $bpk-spacing-xs 0 0;
  }

  &__duration {
    display: flex;
    justify-content: flex-end;
    color: $bpk-color-gray-500;
  }

  &__stops {
    display: flex;
    justify-content: flex-end;
    position: relative;
    top: .375rem;
  }

  &--direct {
    color: $bpk-color-green-500;
  }

  &--with-stops {
    color: $bpk-color-red-500;
  }

  &__departure, &__arrival {
    @include bpk-text-base;
    display: flex;
    flex-direction: column;
    padding: 0 $bpk-spacing-sm;
  }

  &__time {
    color: $bpk-color-gray-900;
  }

  &__place {
    color: $bpk-color-gray-500;
  }

  &__arrow-container {
    padding: $bpk-spacing-sm $bpk-spacing-xs;
  }

  &__arrow {
    width: .875rem;
    height: .875rem;
  }

  &__airline {
    @include bpk-border-radius-xs;
    width: $bpk-icon-size-lg;
    height: $bpk-icon-size-lg;
    position: relative;
    top: -0.375rem;
  }
}
