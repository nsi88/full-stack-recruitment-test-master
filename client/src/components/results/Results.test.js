import React from 'react';
import Results from './Results';
import renderer from 'react-test-renderer';
import serverResponse from '../../../../test/fixtures/server-response.js'

test('spinner render', () => {
  const tree = renderer
    .create(<Results fetchingInProgress={true} error={null} results={[]}/>)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test('error render', () => {
  const tree = renderer
    .create(<Results fetchingInProgress={false} error="Something went wrong." results={[]}/>)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test('results render', () => {
  const tree = renderer
    .create(<Results fetchingInProgress={false} error={null} results={serverResponse.Itineraries}/>)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
