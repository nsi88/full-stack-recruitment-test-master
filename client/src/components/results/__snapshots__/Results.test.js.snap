// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`error render 1`] = `
<div
  className="error"
>
  <div
    className={null}
    onTransitionEnd={[Function]}
    style={
      Object {
        "MozTransition": "height 200ms ease ",
        "OTransition": "height 200ms ease ",
        "WebkitTransition": "height 200ms ease ",
        "height": "auto",
        "msTransition": "height 200ms ease ",
        "overflow": "visible",
        "transition": "height 200ms ease ",
      }
    }
  >
    <div>
      <span
        onTransitionEnd={[Function]}
      >
        <section
          className="bpk-banner-alert bpk-banner-alert--error"
        >
          <header
            className="bpk-banner-alert__header"
            onClick={[Function]}
            role="alert"
          >
            <span
              className="bpk-banner-alert__icon"
            >
              <span
                style={
                  Object {
                    "display": "inline-block",
                    "lineHeight": "1.125rem",
                    "marginTop": "0.1875rem",
                    "verticalAlign": "top",
                  }
                }
              >
                <svg
                  className="bpk-banner-alert__error-icon"
                  height="18"
                  style={
                    Object {
                      "height": "1.125rem",
                      "width": "1.125rem",
                    }
                  }
                  viewBox="0 0 24 24"
                  width="18"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M12 2C6.5 2 2 6.5 2 12s4.5 10 10 10 10-4.5 10-10S17.5 2 12 2zm1.5 16.1c0 .5-.4.9-.9.9h-1.1c-.5 0-.9-.4-.9-.9V11c0-.5.4-.9.9-.9h1.1c.5 0 .9.4.9.9v7.1zm0-11.1c0 .5-.5 1-1 1h-1c-.5 0-1-.5-1-1V6c0-.5.5-1 1-1h1c.5 0 1 .5 1 1v1z"
                  />
                </svg>
              </span>
            </span>
             
            <span
              className="bpk-banner-alert__message"
            >
              Something went wrong.
            </span>
             
          </header>
          <div
            onTransitionEnd={[Function]}
            style={
              Object {
                "MozTransition": "height 200ms ease ",
                "OTransition": "height 200ms ease ",
                "WebkitTransition": "height 200ms ease ",
                "height": 0,
                "msTransition": "height 200ms ease ",
                "overflow": "hidden",
                "transition": "height 200ms ease ",
              }
            }
          >
            <div>
              <div
                className="bpk-banner-alert__children-container"
              />
            </div>
          </div>
        </section>
      </span>
    </div>
  </div>
</div>
`;

exports[`results render 1`] = `
<ul
  className="results"
>
  <li
    className="result"
  >
    <div
      className="result__legs"
    >
      <div
        className="leg"
      >
        <div
          className="leg__main-info"
        >
          <img
            alt="Flybe"
            className="leg__airline"
            src="https://logos.skyscnr.com/images/airlines/favicon/BE.png"
          />
          <div
            className="leg__departure"
          >
            <div
              className="leg__time"
            >
              18:20
            </div>
            <div
              className="leg__place"
            >
              EDI
            </div>
          </div>
          <div
            className="leg__arrow-container"
          >
            <img
              alt="to"
              className="leg__arrow"
              src="test-file-stub"
            />
          </div>
          <div
            className="leg__arrival"
          >
            <div
              className="leg__time"
            >
              20:00
            </div>
            <div
              className="leg__place"
            >
              LHR
            </div>
          </div>
        </div>
        <div
          className="leg__additional-info"
        >
          <div
            className="leg__duration"
          >
            1h 40
          </div>
          <div
            className="leg__stops leg--direct"
          >
            Direct
          </div>
        </div>
      </div>
      <div
        className="leg"
      >
        <div
          className="leg__main-info"
        >
          <img
            alt="Flybe"
            className="leg__airline"
            src="https://logos.skyscnr.com/images/airlines/favicon/BE.png"
          />
          <div
            className="leg__departure"
          >
            <div
              className="leg__time"
            >
              18:20
            </div>
            <div
              className="leg__place"
            >
              LHR
            </div>
          </div>
          <div
            className="leg__arrow-container"
          >
            <img
              alt="to"
              className="leg__arrow"
              src="test-file-stub"
            />
          </div>
          <div
            className="leg__arrival"
          >
            <div
              className="leg__time"
            >
              20:00
            </div>
            <div
              className="leg__place"
            >
              EDI
            </div>
          </div>
        </div>
        <div
          className="leg__additional-info"
        >
          <div
            className="leg__duration"
          >
            1h 40
          </div>
          <div
            className="leg__stops leg--direct"
          >
            Direct
          </div>
        </div>
      </div>
    </div>
    <div
      className="result__bottom"
    >
      <div
        className="result__price-info"
      >
        <div
          className="result__price"
        >
          £
           
          100
        </div>
        <div
          className="result__agent"
        >
          flybe
        </div>
      </div>
      <button
        className="result__select"
      >
        Select
      </button>
    </div>
  </li>
  <li
    className="result"
  >
    <div
      className="result__legs"
    >
      <div
        className="leg"
      >
        <div
          className="leg__main-info"
        >
          <img
            alt="Flybe"
            className="leg__airline"
            src="https://logos.skyscnr.com/images/airlines/favicon/BE.png"
          />
          <div
            className="leg__departure"
          >
            <div
              className="leg__time"
            >
              18:20
            </div>
            <div
              className="leg__place"
            >
              EDI
            </div>
          </div>
          <div
            className="leg__arrow-container"
          >
            <img
              alt="to"
              className="leg__arrow"
              src="test-file-stub"
            />
          </div>
          <div
            className="leg__arrival"
          >
            <div
              className="leg__time"
            >
              20:00
            </div>
            <div
              className="leg__place"
            >
              LHR
            </div>
          </div>
        </div>
        <div
          className="leg__additional-info"
        >
          <div
            className="leg__duration"
          >
            1h 40
          </div>
          <div
            className="leg__stops leg--direct"
          >
            Direct
          </div>
        </div>
      </div>
      <div
        className="leg"
      >
        <div
          className="leg__main-info"
        >
          <img
            alt="Flybe"
            className="leg__airline"
            src="https://logos.skyscnr.com/images/airlines/favicon/BE.png"
          />
          <div
            className="leg__departure"
          >
            <div
              className="leg__time"
            >
              20:45
            </div>
            <div
              className="leg__place"
            >
              LHR
            </div>
          </div>
          <div
            className="leg__arrow-container"
          >
            <img
              alt="to"
              className="leg__arrow"
              src="test-file-stub"
            />
          </div>
          <div
            className="leg__arrival"
          >
            <div
              className="leg__time"
            >
              22:25
            </div>
            <div
              className="leg__place"
            >
              EDI
            </div>
          </div>
        </div>
        <div
          className="leg__additional-info"
        >
          <div
            className="leg__duration"
          >
            1h 40
          </div>
          <div
            className="leg__stops leg--direct"
          >
            Direct
          </div>
        </div>
      </div>
    </div>
    <div
      className="result__bottom"
    >
      <div
        className="result__price-info"
      >
        <div
          className="result__price"
        >
          £
           
          100
        </div>
        <div
          className="result__agent"
        >
          flybe
        </div>
      </div>
      <button
        className="result__select"
      >
        Select
      </button>
    </div>
  </li>
</ul>
`;

exports[`spinner render 1`] = `
<div
  className="spinner"
>
  <div
    className="spinner__icon"
  />
  <div
    className="spinner__text"
  >
    Searching
  </div>
</div>
`;
